### DISCUSSION TITLE
Mumble Client & Server SDK

### DISCUSSION THREAD

**Discussion by pwfran** [mumble-voip/libmumble#17] [Q&A]
Hi,

 we had conversation here  

https://forums.mumble.info/topic/2209-mumble-client-amp-server-sdk/

but now it seems to be archived. I am trying to create my custom voice client using libmumble, based on example client in libmumble-master/examples/ExampleClient directory. Modifying this example client I am able to connect to server and receive voice packets from stock client on another machine, but I cannot decode them and do further processing.
Before the old forum was archived, @davidebeatrici asked for my code - I am attaching modified main.cpp file I use (named main.cpp.txt because it refused to attach .cpp). When I am talking to the remote client, custom client receives UDPTunnel messages (line 140). Then it passes

 if (pack(msg))

condition (line 146) but fails to pass 

if(msg.pack(audio))

(line 149).

So, I get sequence of printed messages:
=== feedback.pack - type: UDPTunnel
= UDPTunnel
### UDPTunnel pack(msg)

Is this some basic mistake ? I would appreciate any support.

Thanks a lot for everyone for any help I got so far and sorry for being annoying - I couldn't find any example of using libmumble to create fully functional client and documentation is  quite limited.

Best regards,
Pawel
[main.cpp.txt](https://github.com/mumble-voip/mumble/files/11781310/main.cpp.txt)

**Comment by davidebeatrici** [comment]
Hi,

Thank you for the recap and sorry about the archived thread, it seems to be missing our latest messages.

Could you check whether the function is returning [here](https://github.com/mumble-voip/libmumble/blob/4c1b92064017cd3b7c2f0e13d0fcdce79479bcfd/src/Pack.cpp#L625-L627), please?

> Thanks a lot for everyone for any help I got so far and sorry for being annoying - I couldn't find any example of using libmumble to create fully functional client and documentation is quite limited.

No worries, documentation is indeed missing (aside from the examples).

**Reply by pwfran** [reply]
Hi,
 sorry for the delay, I was sick for more than a week.
I put printf instructions before, inside and after condition, rebuilt everything and NO - I see only messages before and after the condition, so it does not seem to return here.

**Reply by davidebeatrici** [reply]
Hi,

No worries. Could you also check that the message type is `Type::UDPTunnel`, please?

**Reply by pwfran** [reply]
Hi,

 yes, condition message.type() == Type::UDPTunnel, checked at the beginning of the operator() is true.

**Reply by davidebeatrici** [reply]
Is it perhaps Protobuf's `ParseFromArray()` that is failing? You can check inside the `PARSE_PROTO_MESSAGE` macro.

**Reply by pwfran** [reply]
Yes, I put message printing in 

if (!msg.ParseFromArray(data, static_cast< int >(size)))

condition bedore returning false in PARSE_PROTO_MESSAGE definition and it endeed is printed when client receives messages.

According to printed messages, it seems that for this fragment in my modified client (main.cpp attached previously):

	else if(type == Type::UDPTunnel) 
        {
            printf("= UDPTunnel\n");
            tcp::Message::UDPTunnel msg;
            udp::Message::Audio audio;

            if (pack(msg)) 
            {
                printf("### UDPTunnel pack(msg)\n");
                if(msg.pack(audio))
                {
                    printf("    >>> Audio\n");
                    const auto nSamples = decoder.packetSamples(audio.opusData);
                    if (nSamples) 
                    {
                        printf("   Samples: %d\n", nSamples);
                        std::vector< float > floatData(nSamples);
                        const auto floatDataWritten = decoder(floatData, audio.opusData);

                        std::vector< int16_t > integerData(nSamples);
                        const auto integerDataWritten = decoder(integerData, audio.opusData);
                    }
                }
            }
        }

the error in PARSE_PROTO_MESSAGE is triggered for call msg.pack(audio). The call pack(msg) is successful, the message "### UDPTunnel pack(msg)\n" is printed.

**Reply by davidebeatrici** [reply]
In that case it's likely that the incoming packet is corrupted... is it sent by a tested client?

**Reply by pwfran** [reply]
It is sent from Linux, OpenSUSE 15.4 - mumble client from built in package (mumble-1.4.230-bp154.1.55.x86_64). Does standard Mumble client use the same format as libmumble or maybe not ?

**Reply by davidebeatrici** [reply]
Oh, sorry for taking so long to realize what the issue is... there is a new format for UDP packets in 1.5.x: mumble-voip/mumble#5594

Backward compatibility is provided in libmumble and the legacy format can be used fairly easily:

```c++
#include <mumble/PacketDataStream.hpp>

auto data = pack.data();
if (data.size() < 1 + 1 + 1) {
	// Audio packets must contain at least:
	//
	// - One header byte.
	// - One byte varint-encoding the sequence number.
	// - One byte of audio payload.
	return;
}

if (legacy::udp::type(data) != legacy::udp::Type::VoiceOpus) {
	return;
}

// const uint32_t targetOrContext = data[0] & 0x1f;

PacketDataStream stream(data.subspan(1));

uint32_t senderSession;
stream >> senderSession;

uint64_t frameNumber;
stream >> frameNumber;

uint64_t header;
stream >> header;

// const bool isLastFrame = header & 0x2000;

const auto opusData = stream.seek().first(header & 0x1fff);
```

**Reply by pwfran** [reply]
Thanks, I put this code inside:

else if(type == Type::UDPTunnel) 
        {
            tcp::Message::UDPTunnel msg;
            udp::Message::Audio audio;

            if (pack(msg)) 
            {
                 // YOUR CODE HERE

           }
      }

And if (legacy::udp::type(data) != legacy::udp::Type::VoiceOpus)  condition fails. 
It seems that legacy::udp::type(data) is 0 while legacy::udp::Type::VoiceOpus is 4.

**Reply by davidebeatrici** [reply]
What's the packet size?

**Reply by pwfran** [reply]
When I say something to the connected peer, my compiled client gets a number of packets, data.size() is usually 105 - but sometimes there are single packets with other sizes: 55 or 104. But 105 dominates the set.
